libafl/src/mutators/string.rs

```diff
@@ -0,0 +1,602 @@
+//! Mutators for preserving string categories, which may be useful for certain targets which are primarily string-oriented.
+use alloc::vec::Vec;
+use core::{
+    cmp::{Ordering, Reverse},
+    ops::Range,
+};
+
+use libafl_bolts::{rands::Rand, Error, HasLen, Named};
+
+use crate::{
+    corpus::{CorpusId, HasTestcase, Testcase},
+    inputs::{BytesInput, HasBytesVec},
+    mutators::{rand_range, MutationResult, Mutator, Tokens},
+    stages::{
+        extract_metadata,
+        mutational::{MutatedTransform, MutatedTransformPost},
+        StringIdentificationMetadata,
+    },
+    state::{HasCorpus, HasMaxSize, HasMetadata, HasRand},
+};
+
+/// Input which contains the context necessary to perform unicode mutations
+pub type UnicodeInput = (BytesInput, StringIdentificationMetadata);
+
+impl<S> MutatedTransform<BytesInput, S> for UnicodeInput
+where
+    S: HasCorpus<Input = BytesInput> + HasTestcase,
+{
+    type Post = StringIdentificationMetadata;
+
+    fn try_transform_from(
+        base: &mut Testcase<BytesInput>,
+        state: &S,
+        _corpus_idx: CorpusId,
+    ) -> Result<Self, Error> {
+        let input = base.load_input(state.corpus())?.clone();
+        let metadata = base.metadata::<StringIdentificationMetadata>().cloned()?;
+        Ok((input, metadata))
+    }
+
+    fn try_transform_into(self, _state: &S) -> Result<(BytesInput, Self::Post), Error> {
+        Ok(self)
+    }
+}
+
+impl<S> MutatedTransformPost<S> for StringIdentificationMetadata
+where
+    S: HasTestcase,
+{
+    fn post_exec(
+        self,
+        state: &mut S,
+        _stage_idx: i32,
+        corpus_idx: Option<CorpusId>,
+    ) -> Result<(), Error> {
+        if let Some(corpus_idx) = corpus_idx {
+            let mut tc = state.testcase_mut(corpus_idx)?;
+            tc.add_metadata(self)
+        }
+        Ok(())
+    }
+}
+
+const MAX_CHARS: usize = 16;
+
+fn choose_start<'a, R: Rand>(
+    rand: &mut R,
+    bytes: &[u8],
+    meta: &'a StringIdentificationMetadata,
+) -> Option<(usize, usize)> {
+    let idx = rand.below(bytes.len() as u64) as usize;
+    let mut options = Vec::new();
+    for (start, range) in meta.ranges() {
```

Comment:
we can do this using interval tree in O(logn) but i'm not sure if it's worth it

Reply:
What would such a solution look like? We can potentially do a performance comparison here.

Reply:
basically we can store this object https://github.com/main--/rust-intervaltree in stages::strings *AND* the list of ranges.

you can build the tree for the input that is in the corpus, and quicken the query "which ranges contain this idx" for this query.
but when you start stacking mutations. it doesn't work. you will change the length of the input during the mutation, and we don't wanna spend O(nlogn) time rebuilding the tree for a testcase that is not in the corpus.

Reply:
Hm, yeah, currently the string analysis stage is linear. Since we're rebuilding this metadata for stacked mutations, I think it's better we stay without interval tree.

Reply:
i mean we can separate extract_metadata() into two functions
extract_metadata_linear() and extract_metadata_tree()

For the input that goes into the corpus we can use extract_metadata_tree(), because this guy takes longer to build but get queried often
For the input generated during stacked mutation we can use extract_metadata_linear()

But keeping the current state is fine :+1: